### PR TITLE
Deprecate LocatableAxes from toolkits

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

`LocatableAxes` in the toolkits provides stuff like `get_axes_locator`, but these are provided in the base `Axes` class. So deprecate all of the classes and related functions. Additionally, remove all code from the deprecated `axes_grid`, and point it to the split apart implementations.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant
- [N/A] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [N/A] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by tacaswell** [comment]
I would rather push this to 3.0 (hoping to tag 2.2RC1 this weekend....only a week late)

**Comment by efiring** [comment]
I haven't looked in detail, but I will be glad to see this go in whenever it does, and the duplication go away.

**Comment by QuLogic** [comment]
Updated deprecation messages to say 3.0 instead of 2.2.

**Comment by QuLogic** [comment]
Rebased due to conflicts. These files are messy, so it would be nice to get this in quicker to reduce conflicts.

**Comment by efiring** [comment]
@anntzer or @dopplershift, perhaps one of you could have a look and merge this if it looks OK, so it doesn't have to be rebased again.

**Comment by anntzer** [comment]
I'm not even close to knowing enough about axes_locator to decide whether the functionality is indeed duplicated or not.

**Comment by QuLogic** [comment]
Mainly you can look at the [`LocatableAxesBase` class](https://github.com/matplotlib/matplotlib/pull/10403/files#diff-443e85aa8793af95778324e69af21dd0L885) which adds `[gs]et_axes_locator` and then modifies `apply_aspect` to use it before drawing. But looking at the regular `_AxesBase.draw`, it [does the same thing](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/axes/_base.py#L2540).

File: lib/mpl_toolkits/axes_grid1/axes_divider.py [hunk 1]
```diff
@@ -882,74 +883,23 @@ def locate(self, nx, ny, nx1=None, ny1=None, axes=None, renderer=None):
...
-
-    def _make_twin_axes(self, *kl, **kwargs):
-        """
-        Need to overload so that twinx/twiny will work with
-        these axes.
-        """
-        if 'sharex' in kwargs and 'sharey' in kwargs:
-            raise ValueError("Twinned Axes may share only one axis.")
-        ax2 = type(self)(self.figure, self.get_position(True), *kl, **kwargs)
-        ax2.set_axes_locator(self.get_axes_locator())
```
Comment (efiring) [inline lib/mpl_toolkits/axes_grid1/axes_divider.py]: The base `Axes._make_twin_axes` is missing this locator-copying line--maybe that is a bug.

**Comment by efiring** [comment]
`LocatableAxesBase` dates all the way back to 8e3b9312, Dec. 9, 2008, when all of the AxesGrid family was started.  That original commit (which looks like a combination of SVN commits?) includes the duplication of `get_axes_locator`, etc., and its use in `draw` prior to `apply_aspect`, in `Axes` and in `LocatableAxesBase`.  It looks to me like originally it was in the latter, and then @leejjoon decided to add it to `Axes` but didn't remove it from `LocatableAxesBase`.  The `_axes_locator` is used only in connection with the AxesGrid family, as far as I can see; there are references to it in tight_layout and bbox_inches=`tight`, but only the AxesGrid family actually sets it.

It looks to me like we need to either support `_axes_locator` more thoroughly in `Axes` (which at the moment means taking care to keep it synchronized in the twin case, requiring a modification of the setter as well as of `_make_twin_axes`), or strip it out of `Axes` and leave the locatable axes class, factory function, etc. in place, possibly with some changes.

I'm not sure which would be best; it's a little odd to have an attribute in `Axes` that is used only by a toolkit, isn't it?  On the other hand, the cost is fairly low.

**Comment by QuLogic** [comment]
There are times when @leejjoon [recommended using `axes_locator`](https://www.mail-archive.com/matplotlib-users@lists.sourceforge.net/msg11228.html) and not the full toolkit, so I think the intention was to put it into the base library, but I can't say that 100%. I can also find a [random use in the wild](https://github.com/SauceCat/PDPbox/blob/master/pdpbox/pdp.py#L1245), but it's hard to search when a bunch of people commit their venvs to git.

**Comment by jklymak** [comment]
Ooh, this fell through the cracks.  Any chance we can resurrect?

**Comment by QuLogic** [comment]
I can rebase; depends on @efiring's thoughts.

**Comment by leejjoon** [comment]
My opinion is highly biased as I am the one who introduced the axes_locator. But I would like to see it stays. While I understand this is underused, this makes certain things a lot easier (although may not relevant for normal users). Also, #11026 requires axes_locator.

**Comment by leejjoon** [comment]
Yes. I was just expressing my opinion to what @efiring said, and I was referring to axes_locator in the base. I am happy with the axes_grid becoming deprecated.

**Comment by efiring** [comment]
I'm more than a little confused--I haven't looked at this for a while, it deals with functionality I am not familiar with, and so the spinup in trying to come back to it is considerable.  @leejjoon, it's good to hear from you--but could you be more explicit with respect to this PR?  Do you favor it as is (after a rebase), or with minor changes, or major changes, or not at all?

**Comment by leejjoon** [comment]
I am in favor of this PR. In fact, I think we can actually remove `axes_grid` toolkit from the code base.
On the other hand, I would prefer `_axes_locator` stays within `Axes`. Let me know if I am still not clear.

**Comment by efiring** [comment]
@leejjoon,  thanks very much, that's what I thought. @QuLogic, let's try to get this in now.  Once it is rebased and passes the tests, anyone is welcome to merge it.

**Comment by QuLogic** [comment]
OK, rebase is done.

**Pull request merged**